doc: improve python intro (#5526)
OSGeo/grass: man/mkdocs/docs/python_intro.md

    - GRASS Development Team
---

# Python introduction

GRASS Python interface provides libraries to create GRASS scripts and access
the internal data structures of GRASS. The Python interface consists of
two main libraries:
*[grass.script](https://grass.osgeo.org/grass-stable/manuals/libpython/script_intro.html)*
provides a Python interface to GRASS tools
and *[grass.pygrass](https://grass.osgeo.org/grass-stable/manuals/libpython/pygrass_index.html)*
enables access to the internal data structures of GRASS.

> ## GRASS Script

Reviewer: ```suggestion
## Scripting API
```

### Setup

> To get started with scripting, you must first append the GRASS
> Python path to the system path and then import the *grass.script* library. From

Reviewer: ```suggestion
To get started with scripting, you must first append the GRASS
Python path to the system path and then import the *grass.script* library
(path append may not be needed on some systems). From
```

import grass.script as gs

# Create a new project
gs.create_project(path="path/to/my_project", epsg="3358")

# Initialize the GRASS session

> with gs.init("path/to/my_project") as session:

Reviewer: ```suggestion
with gs.setup.init("path/to/my_project") as session:
```

    # Run GRASS tools
    gs.run_command("r.import", input="/path/to/elevation.tif", output="elevation")
    gs.run_command("g.region", raster="elevation")
    gs.run_command("r.slope.aspect", elevation="elevation", slope="slope")
```

If you are running a script in an already initialized GRASS session, you

> can run the tools right away:

Reviewer: ```suggestion
can run the tools right away (without the call to __init__):
```

| [gs.raster_what](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.raster.raster_what) | Query raster at coordinates |
| [gs.mapcalc](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.raster.mapcalc) | Execute a raster algebra expression |
| [vector_columns](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.vector.vector_columns) | Get the columns of a vector attribute table                  |
| [vector_info](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.vector.vector_info) | Get information about a vector map |
| [vector_db_select](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.vector.vector_db_select) | Get attribute data of a vector map |
| [vector_info_topo](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.vector.vector_info_topo) | Get topology information about a vector map |
| [mapcalc3d](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.raster3d.mapcalc3d) | Execute a 3D raster algebra expression |

Here is an example how to use these helper functions:

```python

> # Get region resolution to get cell size

Reviewer: ```suggestion
# Get region resolution to get cell size (assuming projected CRS)
```

region = gs.region()
cell_area = region["nsres"] * region["ewres"]

> # Compute volume with raster algebra

Reviewer: ```suggestion
# Compute volume with raster algebra (use area() in real life)
```

gs.mapcalc(f"volume = elevation * {cell_area}")

# Get number of points in a vector map
num_points = gs.vector_info_topo("hospitals")["points"])
```

### NumPy interface

The GRASS Python API includes a NumPy interface that allows you to read
and write raster data as NumPy arrays.
This makes it easy to integrate GRASS with the broader Python scientific
stack for advanced analysis and custom modeling.
Using *[grass.script.array](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.array.array)*
and *[grass.script.array3d](https://grass.osgeo.org/grass-stable/manuals/libpython/script.html#script.array.array3d)*,

> you can load raster maps into memory, perform array-based operations,
> and write results back as a GRASS raster or 3D raster.

Reviewer: The array actually are not in memory, at least not initially. The intro covers only the "jump to NumPy" case but not the "jump to GRASS" case. Here is a suggestion to fix both:

```suggestion
you can switch between GRASS raster maps and NumPy arrays, run GRASS tools,
and perform array-based operations as needed.
It works for rasters as well as for 3D rasters in GRASS.
```

This example shows a workflow for writing a NumPy array
to a GRASS raster, running a GRASS tool, and loading the result as a NumPy array:

```python
import numpy as np
from grass.script import array as garray

# Create a 100x100 sinusoidal elevation surface
xx, yy = np.meshgrid(np.linspace(0, 1, 100), np.linspace(0, 1, 100))
elevation_array = np.sin(xx) + np.cos(yy)

# Set the region to match the array dimensions and resolution
gs.run_command("g.region", n=elevation_array.shape[0], s=0,
               e=elevation_array.shape[1], w=0, res=1)

# Write the NumPy array to a new GRASS raster map
map2d = garray.array()
map2d[:] = elevation_array

# Output the array to a raster map in GRASS

> map2d.write("elevation", overwrite=True)

Reviewer: Confusing "Write" then "Output". Merge these two into one block with one comment.

# Compute e.g., flow accumulation
gs.run_command("r.watershed", elevation="elevation", accumulation="accumulation")

# Load as numpy array
accumulation_array = garray.array("accumulation")
```

This example demonstrates reading an existing GRASS raster into a NumPy array,
modifying the array, and writing the modified array back to a GRASS raster:

```python
import numpy as np
import seaborn as sns
from grass.script import array as garray

# Read elevation as numpy array
elev = garray.array(mapname="elevation")
# Plot raster histogram
sns.histplot(data=elev.ravel(), kde=True)
# Modify values
elev_2 *= 2
# Write modified array into a GRASS raster
elev_2.write(mapname="elevation_2")
```

> ## Object-Oriented GRASS

Reviewer: ```suggestion
## Object-Oriented API
```

Bellow, remove gisdbase, location, and mapset headings, and move Region heading one level up.
